File: platform/macos/Bridge/NativeSecurity.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import TidyVNC

extension tidyvnc_security_choice: ABIValue {}
extension tidyvnc_security_selection: ABIValue {}
public struct NativeSecurityChoice: Sendable, Identifiable {
  public enum Protection: UInt32, CaseIterable, Sendable { case unencrypted = 0, anonymousTLS, x509TLS, rsaAES, rsaAuthentication, legacyAuthentication }
  public enum Credentials: UInt32, Sendable { case none = 0, password, usernamePassword, serverSelected }
  public let id: UInt32, name: String, available: Bool
  public let protection: Protection, credentials: Credentials, aesBits: UInt32
}
func securityText<T>(_ value: T) -> String {
  withUnsafeBytes(of: value) { String(decoding: $0.prefix(while: { $0 != 0 }),as: UTF8.self) }
}
public struct NativeSecuritySelection: Equatable, Sendable {
  public let types: [UInt32]
  public let canonical: String
  // nil is the compiled default; "" explicitly denies all methods.
  public init(_ text: String? = nil) throws {
    guard (text?.utf8.prefix(1025).count ?? 0) <= 1024 else { throw NativeError(.resourceLimit,"Security selection is too long") }
    var result = abi(tidyvnc_security_selection.self)
    _ = try withText(text ?? "") { input in
      try checked { tidyvnc_security_resolve(input,text == nil ? 1 : 0,&result,$0) }
    }
    guard result.count <= 32 else { throw NativeError(.internalFailure,"Invalid security selection count") }
    types = withUnsafeBytes(of: result.types) { Array($0.bindMemory(to: UInt32.self).prefix(Int(result.count))) }
    canonical = securityText(result.canonical)
  }
  public static func choices() throws -> [NativeSecurityChoice] {
    var result: [NativeSecurityChoice] = []
    for index in 0..<32 {
      var value = abi(tidyvnc_security_choice.self)
      if try checked(allowing: [.ok,.noChange], { tidyvnc_security_choice_at(UInt32(index),&value,$0) }) == .noChange { return result }
      guard let protection = NativeSecurityChoice.Protection(rawValue:value.protection),
            let credentials = NativeSecurityChoice.Credentials(rawValue:value.credentials) else { throw NativeError(.unsupported,"Unknown security metadata") }
      result.append(.init(id:value.type,name:securityText(value.name),available:value.available != 0,
        protection:protection,credentials:credentials,aesBits:value.aes_bits))
    }
    throw NativeError(.resourceLimit,"Security catalog exceeds its limit")
  }
}

// GnuTLS may consult library configuration. Storage actors call this preflight;
// view bodies and per-keystroke validity checks must remain free of library I/O.
public enum NativeTLSPriority {
  public static func validate(_ text: String) throws {
    guard text.utf8.prefix(4097).count <= 4096 else { throw NativeError(.resourceLimit,"TLS priority is too long") }
    _ = try withText(text) { input in try checked { tidyvnc_tls_priority_validate(input,$0) } }
  }
}

extension tidyvnc_security_configuration: ABIValue {}
extension tidyvnc_security_update: ABIValue {}
public struct NativeSessionSecurity: Equatable, Sendable {
  public let revision: UInt64, generation: UInt64, editable: Bool
  public let preferences: NativeSecurityPreferences
  public let trustFiles: NativeTrustFiles
}

extension tidyvnc_sharing: ABIValue {}
public struct NativeConnectionOptions: Equatable, Sendable {
  public let shared: Bool, reconnectOnError: Bool, editable: Bool
  public let revision: UInt64, generation: UInt64
  public let sharedSource: NativeOptionSource, reconnectSource: NativeOptionSource
}

File: platform/windows/TidyVNC.Native/Bridge/NativeSecurity.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

public sealed record NativeSecurityChoice(uint Id, string Name, bool Available, NativeSecurityChoice.ProtectionKind Protection,
                                          NativeSecurityChoice.CredentialKind Credentials, uint AesBits)
{
    public enum ProtectionKind : uint { Unencrypted = 0, AnonymousTls, X509Tls, RsaAes, RsaAuthentication, LegacyAuthentication }
    public enum CredentialKind : uint { None = 0, Password, UsernamePassword, ServerSelected }
}

/// <summary>A resolved SecurityTypes selection (NativeSecuritySelection).</summary>
public sealed record NativeSecuritySelection
{
    public IReadOnlyList<uint> Types { get; }
    public string Canonical { get; }

    /// <summary>Null is the compiled default; "" explicitly denies every method.</summary>
    public unsafe NativeSecuritySelection(string? text = null)
    {
        var bytes = AbiText.Utf8(text ?? "");
        if (bytes.Length > 1024) throw new NativeError(NativeStatus.ResourceLimit, "Security selection is too long");
        var result = Abi.Init<tidyvnc_security_selection>();
        var error = Abi.Init<tidyvnc_error>();
        fixed (byte* p = bytes)
            Abi.Check(NativeMethods.tidyvnc_security_resolve(AbiText.Span(p, bytes.Length), text is null ? 1u : 0u, &result, &error), &error);
        if (result.count > 32) throw new NativeError(NativeStatus.InternalFailure, "Invalid security selection count");
        var types = new uint[result.count];
        for (var i = 0; i < types.Length; i++) types[i] = result.types[i];
        Types = types; Canonical = AbiText.Fixed(result.canonical, 1025);
    }

    public bool Equals(NativeSecuritySelection? other) => other is not null && Canonical == other.Canonical && Types.SequenceEqual(other.Types);
    public override int GetHashCode() => Canonical.GetHashCode(StringComparison.Ordinal);

    public static unsafe IReadOnlyList<NativeSecurityChoice> Choices()
    {
        var result = new List<NativeSecurityChoice>();
        var error = Abi.Init<tidyvnc_error>();
        for (uint index = 0; index < 32; index++)
        {
            var value = Abi.Init<tidyvnc_security_choice>();
            if (Abi.Check(NativeMethods.tidyvnc_security_choice_at(index, &value, &error), &error, NativeStatus.Ok, NativeStatus.NoChange) == NativeStatus.NoChange)
                return result;
            if (!Enum.IsDefined((NativeSecurityChoice.ProtectionKind)value.protection) ||
                !Enum.IsDefined((NativeSecurityChoice.CredentialKind)value.credentials))
                throw new NativeError(NativeStatus.Unsupported, "Unknown security metadata");
            result.Add(new NativeSecurityChoice(value.type, AbiText.Fixed(value.name, 32), value.available != 0,
                (NativeSecurityChoice.ProtectionKind)value.protection, (NativeSecurityChoice.CredentialKind)value.credentials, value.aes_bits));
        }
        throw new NativeError(NativeStatus.ResourceLimit, "Security catalog exceeds its limit");
    }
}

/// <summary>GnuTLS priority validation; may read library configuration, so never call it per keystroke on the UI thread.</summary>
public static class NativeTlsPriority
{
    public static unsafe void Validate(string text)
    {
        var bytes = AbiText.Utf8(text);
        if (bytes.Length > 4096) throw new NativeError(NativeStatus.ResourceLimit, "TLS priority is too long");
        var error = Abi.Init<tidyvnc_error>();
        fixed (byte* p = bytes)
            Abi.Check(NativeMethods.tidyvnc_tls_priority_validate(AbiText.Span(p, bytes.Length), &error), &error);
    }
}
